internal/dbunique/db_unique.go

```diff
@@ -32,195 +28,127 @@ var defaultUniqueStates = []rivertype.JobState{ //nolint:gochecknoglobals
-	}
-
-	existing, err := subExec.JobGetByKindAndUniqueProperties(ctx, getParams)
-	if err != nil {
-		if !errors.Is(err, rivertype.ErrNotFound) {
-			return nil, fmt.Errorf("error getting unique job: %w", err)
-		}
-	}
-
-	if existing != nil {
-		// Insert skipped; returns an existing row.
-		return &rivertype.JobInsertResult{Job: existing, UniqueSkippedAsDuplicate: true}, nil
-	}
-
-	jobRow, err := subExec.JobInsertFast(ctx, params)
-	if err != nil {
-		return nil, err
-	}
-
-	if err := subExec.Commit(ctx); err != nil {
-		return nil, err
+func (o *UniqueOpts) StateBitmask() byte {
+	states := defaultUniqueStates
+	if len(o.ByState) > 0 {
+		states = o.ByState
 	}
+	return UniqueStatesToBitmask(states)
+}
 
-	return &rivertype.JobInsertResult{Job: jobRow}, nil
+func UniqueKey(timeGen baseservice.TimeGenerator, uniqueOpts *UniqueOpts, params *riverdriver.JobInsertFastParams) []byte {
+	uniqueKeyString := buildUniqueKeyString(timeGen, uniqueOpts, params)
+	uniqueKeyHash := sha256.Sum256([]byte(uniqueKeyString))
+	return uniqueKeyHash[:]
 }
 
 // Builds a unique key made up of the unique options in place. The key is hashed
 // to become a value for `unique_key` in the fast insertion path, or hashed and
 // used for an advisory lock on the slow insertion path.
-func (i *UniqueInserter) buildUniqueKey(params *riverdriver.JobInsertFastParams, uniqueOpts *UniqueOpts) (string, bool) {
+func buildUniqueKeyString(timeGen baseservice.TimeGenerator, uniqueOpts *UniqueOpts, params *riverdriver.JobInsertFastParams) string {
 	var sb strings.Builder
 
+	if !uniqueOpts.ExcludeKind {
+		sb.WriteString("&kind=" + params.Kind)
+	}
+
 	if uniqueOpts.ByArgs {
+		encodedArgsForUnique := params.EncodedArgs
+
+		{
+			// Get unique JSON keys from the JobArgs struct:
+			uniqueFields, err := getSortedUniqueFieldsCached(params.Args)
+			if err != nil {
+				// Instead of returning an error, we can just proceed using the entire
+				// encoded args.
+				//
+				// TODO: log or how will we find out about this?
+				goto AppendUniqueArgs
+			}
+
+			if len(uniqueFields) == 0 {
+				// Use all encoded args if no unique fields are specified.
+				goto AppendUniqueArgs
+			}
+
+			// Extract unique values from the EncodedArgs JSON
+			uniqueValues := extractUniqueValues(params.EncodedArgs, uniqueFields)
+
+			// Assemble the JSON object using bytes.Buffer
+			// Better to overallocate a bit than to allocate multiple times, so just
+			// assume we'll cap out at the length of the full encoded args.
+			sortedJSONWithOnlyUniqueValues := make([]byte, 0, len(params.EncodedArgs))
+
+			sjsonOpts := &sjson.Options{ReplaceInPlace: true}
+			for i, key := range uniqueFields {
+				if uniqueValues[i] == "undefined" {
+					continue
+				}
+				sortedJSONWithOnlyUniqueValues, err = sjson.SetRawBytesOptions(sortedJSONWithOnlyUniqueValues, key, []byte(uniqueValues[i]), sjsonOpts)
+				if err != nil {
+					// Should not happen unless key was invalid
```

Comment:
The only documented reason for this error to be non-nil is if the json key was invalid.